pandas/core/frame.py

```diff
@@ -4527,8 +4527,11 @@ def _sanitize_column(self, value) -> ArrayLike:
         """
         self._ensure_valid_index(value)
 
-        # We should never get here with DataFrame value
-        if isinstance(value, Series):
+        # We can get there through isetitem with a DataFrame
+        # or through loc single_block_path
+        if isinstance(value, DataFrame):
+            return _reindex_for_setitem(value, self.index)
+        elif isinstance(value, Series):
```

Comment:
This looks a bit weird now, but is closer to main than combining this two branches

Reply:
right so the comment `# We should never get here with DataFrame value` was not true?

the new comment refers to `isetitem` which is not on 1.4.x so slightly confusing.

it looks weird because could have just done `if isinstance(value, (Series, DataFrame)):` instead?

and the case of handling a dict-like value for the rhs was not working on 1.3.5 so we leaving that behavior unchanged here?

and slightly OT for the backport, on main we get a rogue warning?

```
FutureWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  result.loc[:, "var"] = rhs
```

Reply:
> right so the comment # We should never get here with DataFrame value was not true?

Yes, this was incorrect

> the new comment refers to isetitem which is not on 1.4.x so slightly confusing.

good point forgot about this. will remove

> it looks weird because could have just done if isinstance(value, (Series, DataFrame)): instead?

wanted to keep it closer to main, but can combine 

> and the case of handling a dict-like value for the rhs was not working on 1.3.5 so we leaving that behavior unchanged here?

correct

Yeah the warning is a bit annoying. Intend to look into this soonish